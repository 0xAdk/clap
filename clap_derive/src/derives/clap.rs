// Copyright 2018 Guillaume Pinot (@TeXitoi) <texitoi@texitoi.eu>,
// Kevin Knapp (@kbknapp) <kbknapp@gmail.com>, and
// Andrew Hobden (@hoverbear) <andrew@hoverbear.org>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
// This work was derived from Structopt (https://github.com/TeXitoi/structopt)
// commit#ea76fa1b1b273e65e3b0b1046643715b49bec51f which is licensed under the
// MIT/Apache 2.0 license.

use super::{dummies, from_argmatches, into_app, subcommand};
use proc_macro2::Ident;
use proc_macro_error::abort_call_site;
use quote::quote;
use syn::{self, punctuated, token, Attribute, DataEnum};

pub fn derive_clap(input: &syn::DeriveInput) -> proc_macro2::TokenStream {
    use syn::Data::*;

    let ident = &input.ident;

    match input.data {
        Struct(syn::DataStruct {
            fields: syn::Fields::Named(ref fields),
            ..
        }) => {
            dummies::clap_struct(ident);
            gen_for_struct(ident, &fields.named, &input.attrs)
        }
        Struct(syn::DataStruct {
            fields: syn::Fields::Unit,
            ..
        }) => {
            dummies::clap_struct(ident);
            gen_for_struct(
                ident,
                &punctuated::Punctuated::<syn::Field, token::Comma>::new(),
                &input.attrs,
            )
        }
        Enum(ref e) => {
            dummies::clap_enum(ident);
            gen_for_enum(ident, &input.attrs, e)
        }
        _ => abort_call_site!("`#[derive(Clap)]` only supports non-tuple structs and enums"),
    }
}

fn gen_for_struct(
    name: &syn::Ident,
    fields: &punctuated::Punctuated<syn::Field, token::Comma>,
    attrs: &[syn::Attribute],
) -> proc_macro2::TokenStream {
    let (into_app, attrs) = into_app::gen_for_struct(name, fields, attrs);
    let from_arg_matches = from_argmatches::gen_for_struct(name, fields, &attrs);

    quote! {
        impl ::clap::Clap for #name {}

        #into_app
        #from_arg_matches
    }
}

fn gen_for_enum(name: &Ident, attrs: &[Attribute], e: &DataEnum) -> proc_macro2::TokenStream {
    let into_app = into_app::gen_for_enum(name);
    let from_arg_matches = from_argmatches::gen_for_enum(name);
    let subcommand = subcommand::gen_for_enum(name, attrs, e);
    quote! {
        impl ::clap::Clap for #name { }

        #into_app
        #from_arg_matches
        #subcommand
    }
}
