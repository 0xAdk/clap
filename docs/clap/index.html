<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="rustdoc">
    <meta name="description" content="API documentation for the Rust `clap` crate.">
    <meta name="keywords" content="rust, rustlang, rust-lang, clap">

    <title>clap - Rust</title>

    <link rel="stylesheet" type="text/css" href="../main.css">

    
    
</head>
<body class="rustdoc">
    <!--[if lte IE 8]>
    <div class="warning">
        This old browser is unsupported and will most likely display funky
        things.
    </div>
    <![endif]-->

    

    <section class="sidebar">
        
        <p class='location'></p><script>window.sidebarCurrent = {name: 'clap', ty: 'mod', relpath: '../'};</script>
    </section>

    <nav class="sub">
        <form class="search-form js-only">
            <div class="search-container">
                <input class="search-input" name="search"
                       autocomplete="off"
                       placeholder="Click or press 'S' to search, '?' for more options..."
                       type="search">
            </div>
        </form>
    </nav>

    <section id='main' class="content mod">
<h1 class='fqn'><span class='in-band'>Crate <a class='mod' href=''>clap</a><wbr></span><span class='out-of-band'><a href='stability.html'>[stability]</a> <span id='render-detail'>
            <a id="collapse-all" href="#">[-]</a>&nbsp;<a id="expand-all" href="#">[+]</a>
        </span><a id='src-0' href='../src/clap/lib.rs.html#1-155'>[src]</a></span></h1>
<div class='docblock'><p>A simply library for parsing command line arguments when writing
 command line and console applications.</p>

<p>You can use <code>clap</code> to lay out a list of possible valid command line arguments and let <code>clap</code> parse the string given by the user at runtime.
 When using <code>clap</code> you define a set of parameters and rules for your arguments and at runtime <code>clap</code> will determine their validity.
 Also, <code>clap</code> provides the traditional version and help switches &#39;for free&#39; by parsing the list of possible valid arguments lazily at runtime.
 i.e. only when it&#39;s been determined that the user wants or needs to see the help and version information.</p>

<p>After defining a list of possible valid arguments you get a list of matches that the user supplied at runtime. You can then use this list to
 determine the functioning of your program.</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
 <span class='kw'>use</span> <span class='ident'>clap</span>::{<span class='ident'>Arg</span>, <span class='ident'>App</span>, <span class='ident'>SubCommand</span>};

 <span class='comment'>// ...</span>
 
 <span class='kw'>let</span> <span class='ident'>matches</span> <span class='op'>=</span> <span class='ident'>App</span>::<span class='ident'>new</span>(<span class='string'>&quot;MyApp&quot;</span>)
                        .<span class='ident'>version</span>(<span class='string'>&quot;1.0&quot;</span>)
                        .<span class='ident'>author</span>(<span class='string'>&quot;Kevin K. &lt;kbknapp@gmail.com&gt;&quot;</span>)
                        .<span class='ident'>about</span>(<span class='string'>&quot;Does awesome things&quot;</span>)
                        .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;config&quot;</span>)
                                    .<span class='ident'>short</span>(<span class='string'>&quot;c&quot;</span>)
                                    .<span class='ident'>long</span>(<span class='string'>&quot;config&quot;</span>)
                                    .<span class='ident'>help</span>(<span class='string'>&quot;Sets a custom config file&quot;</span>)
                                    .<span class='ident'>takes_value</span>(<span class='boolval'>true</span>))
                        .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;output&quot;</span>)
                                    .<span class='ident'>help</span>(<span class='string'>&quot;Sets an optional output file&quot;</span>)
                                    .<span class='ident'>index</span>(<span class='number'>1</span>))
                        .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;debug&quot;</span>)
                                    .<span class='ident'>short</span>(<span class='string'>&quot;d&quot;</span>)
                                .<span class='ident'>multiple</span>(<span class='boolval'>true</span>)
                                    .<span class='ident'>help</span>(<span class='string'>&quot;Turn debugging information on&quot;</span>))
                        .<span class='ident'>subcommand</span>(<span class='ident'>SubCommand</span>::<span class='ident'>new</span>(<span class='string'>&quot;test&quot;</span>)
                                                .<span class='ident'>about</span>(<span class='string'>&quot;Has test sub functionality&quot;</span>)
                                                .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;verbose&quot;</span>)
                                                            .<span class='ident'>short</span>(<span class='string'>&quot;v&quot;</span>)
                                                            .<span class='ident'>help</span>(<span class='string'>&quot;Display verbose information&quot;</span>)))
                        .<span class='ident'>get_matches</span>();

    <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='ident'>o</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>value_of</span>(<span class='string'>&quot;output&quot;</span>) {
        <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Value for output: {}&quot;</span>, <span class='ident'>o</span>);
    }
 
    <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='ident'>c</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>value_of</span>(<span class='string'>&quot;config&quot;</span>) {
        <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Value for config: {}&quot;</span>, <span class='ident'>c</span>);
    }

 <span class='kw'>match</span> <span class='ident'>matches</span>.<span class='ident'>occurrences_of</span>(<span class='string'>&quot;debug&quot;</span>) {
    <span class='number'>0</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is off&quot;</span>),
        <span class='number'>1</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is kind of on&quot;</span>),
        <span class='number'>2</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is on&quot;</span>),
        <span class='number'>3</span> <span class='op'>|</span> _ <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Don&#39;t be crazy&quot;</span>),
 }
 
 <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='kw-2'>ref</span> <span class='ident'>matches</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>subcommand_matches</span>(<span class='string'>&quot;test&quot;</span>) {
    <span class='kw'>if</span> <span class='ident'>matches</span>.<span class='ident'>is_present</span>(<span class='string'>&quot;verbose&quot;</span>) {
            <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Printing verbose test info...&quot;</span>);
        } <span class='kw'>else</span> {
            <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Not printing regular test info...&quot;</span>);
        }
    }

 <span class='comment'>// more porgram logic goes here...</span>
</pre>

<p>If you were to compile the above program and run it with the flag <code>--help</code> or <code>-h</code> the following output woud be presented</p>

<pre><code class="language-sh"> $ myprog --help
 MyApp 1.0
 Kevin K. &lt;kbknapp@gmail.com&gt;
 Does awesome things
 
 USAGE:
    MyApp [FLAGS] [OPTIONS] [POSITIONAL] [SUBCOMMANDS]
 
 FLAGS:
    -d              Turn debugging information on
    -h,--help       Prints this message
    -v,--version    Prints version information
 
 OPTIONS:
    -c,--config &lt;config&gt;        Sets a custom config file

 POSITIONAL ARGUMENTS:
    output          Sets an optional output file

 SUBCOMMANDS:
    help            Prints this message
        test            Has test sub-functionality
</code></pre>
</div><h2 id='structs' class='section-header'><a href="#structs">Structs</a></h2>
<table>
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.App.html'
                               title='clap::App'>App</a></td>
                        <td class='docblock short'><p>Used to create a representation of the program and all possible command line arguments
 for parsing at runtime.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.Arg.html'
                               title='clap::Arg'>Arg</a></td>
                        <td class='docblock short'><p>The abstract representation of a command line argument used by the consumer of the library.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.ArgMatches.html'
                               title='clap::ArgMatches'>ArgMatches</a></td>
                        <td class='docblock short'><p>Used to get information about the arguments that
where supplied to the program at runtime.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.SubCommand.html'
                               title='clap::SubCommand'>SubCommand</a></td>
                        <td class='docblock short'><p>The abstract representation of a command line subcommand used by the consumer of the library.</p>
</td>
                    </tr>
                </table></section>
    <section id='search' class="content hidden"></section>

    <section class="footer"></section>

    <div id="help" class="hidden">
        <div class="shortcuts">
            <h1>Keyboard shortcuts</h1>
            <dl>
                <dt>?</dt>
                <dd>Show this help dialog</dd>
                <dt>S</dt>
                <dd>Focus the search field</dd>
                <dt>&larrb;</dt>
                <dd>Move up in search results</dd>
                <dt>&rarrb;</dt>
                <dd>Move down in search results</dd>
                <dt>&#9166;</dt>
                <dd>Go to active search result</dd>
            </dl>
        </div>
        <div class="infos">
            <h1>Search tricks</h1>
            <p>
                Prefix searches with a type followed by a colon (e.g.
                <code>fn:</code>) to restrict the search to a given type.
            </p>
            <p>
                Accepted types are: <code>fn</code>, <code>mod</code>,
                <code>struct</code>, <code>enum</code>,
                <code>trait</code>, <code>typedef</code> (or
                <code>tdef</code>).
            </p>
        </div>
    </div>

    

    <script>
        window.rootPath = "../";
        window.currentCrate = "clap";
        window.playgroundUrl = "";
    </script>
    <script src="../jquery.js"></script>
    <script src="../main.js"></script>
    
    <script async src="../search-index.js"></script>
</body>
</html>