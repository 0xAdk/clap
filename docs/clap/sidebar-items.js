initSidebarItems({"struct":[["App","Used to create a representation of the program and all possible command line arguments for parsing at runtime."],["Arg","The abstract representation of a command line argument used by the consumer of the library."],["ArgMatches","Used to get information about the arguments that where supplied to the program at runtime."],["SubCommand","The abstract representation of a command line subcommand used by the consumer of the library."]]});