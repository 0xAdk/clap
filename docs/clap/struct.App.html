<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="rustdoc">
    <meta name="description" content="API documentation for the Rust `App` struct in crate `clap`.">
    <meta name="keywords" content="rust, rustlang, rust-lang, App">

    <title>clap::App - Rust</title>

    <link rel="stylesheet" type="text/css" href="../main.css">

    
    
</head>
<body class="rustdoc">
    <!--[if lte IE 8]>
    <div class="warning">
        This old browser is unsupported and will most likely display funky
        things.
    </div>
    <![endif]-->

    

    <section class="sidebar">
        
        <p class='location'><a href='index.html'>clap</a></p><div class='block struct'><h2>Structs</h2><a class='struct current' href='struct.App.html' title='Used to create a representation of the program and all possible command line arguments  for parsing at runtime.'>App</a><a class='struct ' href='struct.Arg.html' title='The abstract representation of a command line argument used by the consumer of the library.  '>Arg</a><a class='struct ' href='struct.ArgMatches.html' title='Used to get information about the arguments that where supplied to the program at runtime.'>ArgMatches</a></div>
    </section>

    <nav class="sub">
        <form class="search-form js-only">
            <div class="search-container">
                <input class="search-input" name="search"
                       autocomplete="off"
                       placeholder="Click or press 'S' to search, '?' for more options..."
                       type="search">
            </div>
        </form>
    </nav>

    <section id='main' class="content struct">
<h1 class='fqn'><span class='in-band'>Struct <a href='index.html'>clap</a>::<wbr><a class='struct' href=''>App</a><wbr></span><span class='out-of-band'><span id='render-detail'>
            <a id="collapse-all" href="#">[-]</a>&nbsp;<a id="expand-all" href="#">[+]</a>
        </span><a id='src-19' href='../src/clap/app.rs.html#37-59'>[src]</a></span></h1>
<pre class='rust struct'>pub struct App {
    pub name: &amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>,
    pub author: <a class='enum' href='http://doc.rust-lang.org/nightly/core/option/enum.Option.html' title='core::option::Option'>Option</a>&lt;&amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>&gt;,
    pub version: <a class='enum' href='http://doc.rust-lang.org/nightly/core/option/enum.Option.html' title='core::option::Option'>Option</a>&lt;&amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>&gt;,
    pub about: <a class='enum' href='http://doc.rust-lang.org/nightly/core/option/enum.Option.html' title='core::option::Option'>Option</a>&lt;&amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>&gt;,
    // some fields omitted
}</pre><div class='docblock'><p>Used to create a representation of the program and all possible command line arguments
 for parsing at runtime.</p>

<p>Stores a list of all posisble arguments, as well as information displayed to the user such as
 help and versioning information.</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
 <span class='kw'>let</span> <span class='ident'>myprog</span> <span class='op'>=</span> <span class='ident'>App</span>::<span class='ident'>new</span>(<span class='string'>&quot;myprog&quot;</span>)
                   .<span class='ident'>author</span>(<span class='string'>&quot;Me, me@mail.com&quot;</span>)
                      .<span class='ident'>version</span>(<span class='string'>&quot;1.0.2&quot;</span>)
                   .<span class='ident'>about</span>(<span class='string'>&quot;Explains in brief what the program does&quot;</span>)
                   .<span class='ident'>arg</span>(
                            <span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;in_file&quot;</span>).<span class='ident'>index</span>(<span class='number'>1</span>)
                        <span class='comment'>// Add other possible command line argument options here...</span>
                    )
                   .<span class='ident'>get_matches</span>();

 <span class='comment'>// Your pogram logic starts here...</span>
</pre>
</div><h2 class='fields'>Fields</h2>
<table><tr><td id='structfield.name'><a class='stability Unmarked' title='No stability level'></a><code>name</code></td><td><div class='docblock'><p>The name displayed to the user when showing version and help/usage information</p>
</div></td></tr><tr><td id='structfield.author'><a class='stability Unmarked' title='No stability level'></a><code>author</code></td><td><div class='docblock'><p>A string of author(s) if desired. Displayed when showing help/usage information</p>
</div></td></tr><tr><td id='structfield.version'><a class='stability Unmarked' title='No stability level'></a><code>version</code></td><td><div class='docblock'><p>The version displayed to the user</p>
</div></td></tr><tr><td id='structfield.about'><a class='stability Unmarked' title='No stability level'></a><code>about</code></td><td><div class='docblock'><p>A brief explaination of the program that gets displayed to the user when shown help/usage information</p>
</div></td></tr></table><h2 id='methods'>Methods</h2><h3 class='impl'><a class='stability Unmarked' title='No stability level'></a><code>impl <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h3><div class='impl-items'><h4 id='method.new' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.new' class='fnname'>new</a>(n: &amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>) -&gt; <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Creates a new instance of an application requiring a name (such as the binary). Will be displayed
to the user when they print version or help and usage information.</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
<span class='kw'>let</span> <span class='ident'>prog</span> <span class='op'>=</span> <span class='ident'>App</span>::<span class='ident'>new</span>(<span class='string'>&quot;myprog&quot;</span>)
</pre>
</div><h4 id='method.author' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.author' class='fnname'>author</a>(&amp;mut self, a: &amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>) -&gt; &amp;mut <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Sets a string of author(s)</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
.<span class='ident'>author</span>(<span class='string'>&quot;Kevin &lt;kbknapp@gmail.com&gt;&quot;</span>)
</pre>
</div><h4 id='method.about' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.about' class='fnname'>about</a>(&amp;mut self, a: &amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>) -&gt; &amp;mut <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Sets a string briefly describing what the program does</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
.<span class='ident'>about</span>(<span class='string'>&quot;Does really amazing things to great people&quot;</span>)
</pre>
</div><h4 id='method.version' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.version' class='fnname'>version</a>(&amp;mut self, v: &amp;'static <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>) -&gt; &amp;mut <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Sets a string of the version number</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
.<span class='ident'>version</span>(<span class='string'>&quot;v0.1.24&quot;</span>)
</pre>
</div><h4 id='method.arg' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.arg' class='fnname'>arg</a>(&amp;mut self, a: &amp;<a class='struct' href='../clap/struct.Arg.html' title='clap::Arg'>Arg</a>) -&gt; &amp;mut <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Adds an argument to the list of valid possibilties</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
 .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;config&quot;</span>)
                .<span class='ident'>short</span>(<span class='string'>&quot;c&quot;</span>)
            <span class='comment'>// Additional argument configuration goes here...</span>
 )
</pre>
</div><h4 id='method.args' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.args' class='fnname'>args</a>(&amp;mut self, args: <a class='struct' href='http://doc.rust-lang.org/nightly/collections/vec/struct.Vec.html' title='collections::vec::Vec'>Vec</a>&lt;&amp;<a class='struct' href='../clap/struct.Arg.html' title='clap::Arg'>Arg</a>&gt;) -&gt; &amp;mut <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a></code></h4>
<div class='docblock'><p>Adds arguments to the list of valid possibilties</p>

<p>Example:</p>
<pre id='rust-example-rendered' class='rust '>
 .<span class='ident'>args</span>( <span class='macro'>vec</span><span class='macro'>!</span>[<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;config&quot;</span>).<span class='ident'>short</span>(<span class='string'>&quot;c&quot;</span>),
                <span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;debug&quot;</span>).<span class='ident'>short</span>(<span class='string'>&quot;d&quot;</span>)])
</pre>
</div><h4 id='method.get_matches' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.get_matches' class='fnname'>get_matches</a>(&amp;mut self) -&gt; <a class='struct' href='../clap/struct.ArgMatches.html' title='clap::ArgMatches'>ArgMatches</a></code></h4>
</div></section>
    <section id='search' class="content hidden"></section>

    <section class="footer"></section>

    <div id="help" class="hidden">
        <div class="shortcuts">
            <h1>Keyboard shortcuts</h1>
            <dl>
                <dt>?</dt>
                <dd>Show this help dialog</dd>
                <dt>S</dt>
                <dd>Focus the search field</dd>
                <dt>&larrb;</dt>
                <dd>Move up in search results</dd>
                <dt>&rarrb;</dt>
                <dd>Move down in search results</dd>
                <dt>&#9166;</dt>
                <dd>Go to active search result</dd>
            </dl>
        </div>
        <div class="infos">
            <h1>Search tricks</h1>
            <p>
                Prefix searches with a type followed by a colon (e.g.
                <code>fn:</code>) to restrict the search to a given type.
            </p>
            <p>
                Accepted types are: <code>fn</code>, <code>mod</code>,
                <code>struct</code>, <code>enum</code>,
                <code>trait</code>, <code>typedef</code> (or
                <code>tdef</code>).
            </p>
        </div>
    </div>

    

    <script>
        window.rootPath = "../";
        window.currentCrate = "clap";
        window.playgroundUrl = "";
    </script>
    <script src="../jquery.js"></script>
    <script src="../main.js"></script>
    
    <script async src="../search-index.js"></script>
</body>
</html>