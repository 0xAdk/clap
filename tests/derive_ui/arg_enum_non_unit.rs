use clap::ArgEnum;

#[derive(ArgEnum, Clone, Debug)]
enum Opt {
    Foo(usize),
}

fn main() {
    println!("{:?}", Opt::Foo(42));
}
