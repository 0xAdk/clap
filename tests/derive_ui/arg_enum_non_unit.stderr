error: `#[derive(ArgEnum)]` only supports non-unit variants
 --> tests/derive_ui/arg_enum_non_unit.rs:3:10
  |
3 | #[derive(ArgEnum, Clone, Debug)]
  |          ^^^^^^^
  |
  = note: this error originates in the derive macro `ArgEnum` (in Nightly builds, run with -Z macro-backtrace for more info)
